### PR TITLE
Add support for /32 (/128) networks and perNodeBlockSize 1

### PR DESCRIPTION
**Pull request by ykulazhenkov**
Add support for single IP ranges/subnets in all places:

<details>
  <summary>IPPool perNodeBlockSize 1</summary>

```
apiVersion: nv-ipam.nvidia.com/v1alpha1
kind: IPPool
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"nv-ipam.nvidia.com/v1alpha1","kind":"IPPool","metadata":{"annotations":{},"name":"pool1","namespace":"kube-system"},"spec":{"gateway":"192.168.0.3","perNodeBlockSize":1,"subnet":"192.168.0.0/24"}}
  creationTimestamp: "2024-10-14T16:44:59Z"
  generation: 15
  name: pool1
  namespace: kube-system
  resourceVersion: "16150"
  uid: b43c059a-9364-4fbc-b945-1a1f017ddb1b
spec:
  gateway: 192.168.0.3
  perNodeBlockSize: 1
  subnet: 192.168.0.0/24
status:
  allocations:
  - endIP: 192.168.0.2
    nodeName: kind-control-plane
    startIP: 192.168.0.2
  - endIP: 192.168.0.1
    nodeName: kind-worker
    startIP: 192.168.0.1
  - endIP: 192.168.0.4
    nodeName: kind-worker2
    startIP: 192.168.0.4
  - endIP: 192.168.0.5
    nodeName: kind-worker3
    startIP: 192.168.0.5
```
</details>
<details>
  <summary>IPPool /32</summary>

```
apiVersion: nv-ipam.nvidia.com/v1alpha1
kind: IPPool
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"nv-ipam.nvidia.com/v1alpha1","kind":"IPPool","metadata":{"annotations":{},"name":"pool1","namespace":"kube-system"},"spec":{"perNodeBlockSize":1,"subnet":"192.168.1.1/32"}}
  creationTimestamp: "2024-10-14T16:44:59Z"
  generation: 16
  name: pool1
  namespace: kube-system
  resourceVersion: "16441"
  uid: b43c059a-9364-4fbc-b945-1a1f017ddb1b
spec:
  perNodeBlockSize: 1
  subnet: 192.168.1.1/32
status:
  allocations:
  - endIP: 192.168.1.1
    nodeName: kind-control-plane
    startIP: 192.168.1.1
```
</details>

</details>
<details>
  <summary>IPPool /31</summary>

```
apiVersion: nv-ipam.nvidia.com/v1alpha1
kind: IPPool
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"nv-ipam.nvidia.com/v1alpha1","kind":"IPPool","metadata":{"annotations":{},"name":"pool1","namespace":"kube-system"},"spec":{"perNodeBlockSize":1,"subnet":"192.168.1.1/31"}}
  creationTimestamp: "2024-10-14T16:44:59Z"
  generation: 17
  name: pool1
  namespace: kube-system
  resourceVersion: "16586"
  uid: b43c059a-9364-4fbc-b945-1a1f017ddb1b
spec:
  perNodeBlockSize: 1
  subnet: 192.168.1.1/31
status:
  allocations:
  - endIP: 192.168.1.1
    nodeName: kind-control-plane
    startIP: 192.168.1.1
  - endIP: 192.168.1.0
    nodeName: kind-worker
    startIP: 192.168.1.0
```
</details>

</details>
<details>
  <summary>CIDRPool /32</summary>

```
apiVersion: nv-ipam.nvidia.com/v1alpha1
kind: CIDRPool
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"nv-ipam.nvidia.com/v1alpha1","kind":"CIDRPool","metadata":{"annotations":{},"name":"pool1","namespace":"kube-system"},"spec":{"cidr":"192.168.100.0/24","perNodeNetworkPrefix":32,"staticAllocations":[{"prefix":"192.168.100.0/32"}]}}
  creationTimestamp: "2024-10-14T16:08:21Z"
  generation: 9
  name: pool1
  namespace: kube-system
  resourceVersion: "17155"
  uid: 1a90011c-e373-4ed7-bfd9-8bf1d9935a76
spec:
  cidr: 192.168.100.0/24
  perNodeNetworkPrefix: 32
  staticAllocations:
  - prefix: 192.168.100.0/32
status:
  allocations:
  - nodeName: kind-control-plane
    prefix: 192.168.100.1/32
  - nodeName: kind-worker
    prefix: 192.168.100.2/32
  - nodeName: kind-worker2
    prefix: 192.168.100.3/32
  - nodeName: kind-worker3
    prefix: 192.168.100.4/32
```
</details>

<details>
  <summary>CIDRPool /128</summary>

```
apiVersion: nv-ipam.nvidia.com/v1alpha1
kind: CIDRPool
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"nv-ipam.nvidia.com/v1alpha1","kind":"CIDRPool","metadata":{"annotations":{},"name":"pool1","namespace":"kube-system"},"spec":{"cidr":"fd52:2eb5:44::100/128","perNodeNetworkPrefix":128}}
  creationTimestamp: "2024-10-14T16:08:21Z"
  generation: 11
  name: pool1
  namespace: kube-system
  resourceVersion: "17745"
  uid: 1a90011c-e373-4ed7-bfd9-8bf1d9935a76
spec:
  cidr: fd52:2eb5:44::100/128
  perNodeNetworkPrefix: 128
status:
  allocations:
  - nodeName: kind-control-plane
    prefix: fd52:2eb5:44::100/128
```
</details>

`perNodeBlockSize: 1` - useful when you need to deploy one and only one instance of a service per node and you need to make sure that this instance is always have the same predictable IP.

`CIDRPool /32` - useful to assign IPs for loopback interfaces on routers

### CODE REVIEW

File: pkg/ipam-controller/allocator/allocator.go [hunk 1]
```diff
@@ -147,29 +157,44 @@ func (pa *PoolAllocator) checkAllocation(allocRange AllocatedRange) error {
...
 }
 
 // return slice with allocated ranges.
 // ranges are not overlap and are sorted, but there can be "holes" between ranges
 func (pa *PoolAllocator) getAllocationsAsSlice() []AllocatedRange {
-	allocatedRanges := make([]AllocatedRange, 0, len(pa.allocations))
+	allocatedRanges := make([]AllocatedRange, 0, len(pa.allocations)+1)
+
+	if pa.cfg.PerNodeBlockSize == 1 && pa.cfg.Gateway != nil {
+		// in case if perNodeBlockSize is 1 we should not allocated the gateway,
```
Comment (vasrem) [inline pkg/ipam-controller/allocator/allocator.go:193]: nit: `allocate`

File: pkg/ipam-controller/allocator/allocator.go [hunk 2]
```diff
@@ -147,29 +157,44 @@ func (pa *PoolAllocator) checkAllocation(allocRange AllocatedRange) error {
...
+		// add a "virtual" allocatin for the gateway if we detect that only 1 IP is requested per node,
```
Comment (vasrem) [inline pkg/ipam-controller/allocator/allocator.go:194]: nit: `allocation`

File: pkg/ipam-controller/allocator/allocator_test.go [hunk 3]
```diff
@@ -236,6 +236,144 @@ var _ = Describe("Allocator", func() {
...
+			node1Alloc, err := a.AllocateFromPool(ctx, testNodeName1)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(node1Alloc.StartIP.String()).To(BeEquivalentTo("10.10.10.10"))
+			Expect(node1Alloc.EndIP.String()).To(BeEquivalentTo("10.10.10.10"))
+			node2Alloc, err := a.AllocateFromPool(ctx, testNodeName2)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(node2Alloc.StartIP.String()).To(BeEquivalentTo("10.10.10.11"))
+			Expect(node2Alloc.EndIP.String()).To(BeEquivalentTo("10.10.10.11"))
+		})
+		It("load test", func ()  {
```
Comment (vasrem) [inline pkg/ipam-controller/allocator/allocator_test.go:343]: nit: name is misleading, this test is about loading from state from status right? Also, isn't this kinda duplicate of `Load single IP range`?

File: pkg/ipam-node/controllers/cidrpool/cidrpool.go [hunk 4]
```diff
@@ -72,7 +72,7 @@ func (r *CIDRPoolReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
 			}
 			_, nodeSubnet, _ := net.ParseCIDR(alloc.Prefix)
 			startIP := ip.NextIP(nodeSubnet.IP)
-			if ip.IsPointToPointSubnet(nodeSubnet) {
+			if ip.IsPointToPointSubnet(nodeSubnet) || ip.IsSingleIPSubnet(nodeSubnet) {
```
Comment (vasrem) [inline pkg/ipam-node/controllers/cidrpool/cidrpool.go:75]: looks like that we eventually may need some more testing on this package?

**Review by vasrem — APPROVED** [review]
LGTM

A couple of questions and nits

File: pkg/ip/cidr.go [hunk 5]
```diff
@@ -138,6 +138,9 @@ func IsBroadcast(ip net.IP, network *net.IPNet) bool {
 	if network.IP.To4() == nil {
 		return false
 	}
+	if IsPointToPointSubnet(network) || IsSingleIPSubnet(network) {
```
[the remark this answers is not in the document]
Comment (ykulazhenkov) [reply pkg/ip/cidr.go:141]: `IsBroadcast` function was used only for the allocator for IPPools. It was not possible to use /31 in the IPPools before this PR.

File: pkg/ipam-controller/allocator/allocator.go (hunk 1 above)
Comment (ykulazhenkov) [reply pkg/ipam-controller/allocator/allocator.go]: fixed

File: pkg/ipam-controller/allocator/allocator.go (hunk 2 above)
Comment (ykulazhenkov) [reply pkg/ipam-controller/allocator/allocator.go]: fixed

File: pkg/ipam-controller/allocator/allocator_test.go (hunk 3 above)
Comment (ykulazhenkov) [reply pkg/ipam-controller/allocator/allocator_test.go]: yes, they are similar, but they are targeting to test a different things. The first one validates that it is possible to load (from the CR's status) allocation with the single IP. The second test validates that loading is working as expected for /31 and /32 subnets (we have a special handling for this in the code).

I renamed this test to `/31 pool - load allocations`

File: pkg/ipam-node/controllers/cidrpool/cidrpool.go (hunk 4 above)
Comment (ykulazhenkov) [reply pkg/ipam-node/controllers/cidrpool/cidrpool.go:75]: agree, I added some tests to the package

**Comment by ykulazhenkov** [comment]
@vasrem  Thanks for the review! I addressed your comments.

**Review by rollandf — APPROVED** [review]

File: README.md [hunk 6]
```diff
@@ -394,7 +394,7 @@ spec:
 > __Notes:__
 >
 > * pool name is composed of alphanumeric letters separated by dots(`.`) underscores(`_`) or hyphens(`-`).
-> * `perNodeBlockSize` minimum size is 2.
+> * `perNodeBlockSize` minimum size is 1.
```
Comment (adrianchiris) [inline README.md:397]: do you think its worth adding some example usage of /32 or /128 networks ?

**Review by adrianchiris — APPROVED** [review]

**Review by vasrem — APPROVED** [review]

File: README.md (hunk 6 above)
Comment (ykulazhenkov) [reply README.md:397]: This is an edge-case, that is now supported. I don't see any reason to additionally promote such configs. The API is matching all other use-cases.

Comment (adrianchiris) [reply README.md:397]: ack.

**Pull request merged**